inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,187 @@
+
+    # Call oras to pull the image. The password is the JWT token.
+    (out, err_run, code) = PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} "
+                                                   f"--password $ORAS_PASSWORD", env=env)
+    if code != 0:
+        if err_run is not None:
+            raise SotaError("Error to download OTA files with ORAS: " + err_run + ". Code: " + str(code))
+        else:
+            raise SotaError("Error to download OTA files with ORAS. Code: " + str(code))
+    else:
+        dispatcher_broker.telemetry('OTA Download Successful.')
+
+
+def is_enough_space_to_download(manifest_uri: str,
+                                destination_repo: IRepo,
+                                jwt_token: str) -> bool:
+    """Checks if enough free space exists on platform to hold download.
+
+    Calculates the file size from the OCI server and checks if required free space is available on
+    the platform.
+    @param manifest_uri: registry manifest uri
+    @param destination_repo:  desired download destination
+    @param jwt_token: jwt_token provided for access the release server
+    """
+    try:
+        logger.debug(f"Checking OCI artifact size with manifest uri: {manifest_uri}")
+        headers = {
+            "Authorization": f"Bearer {jwt_token}",
+            "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json"
+        }
+        response = requests.get(manifest_uri, headers=headers)
+        if response.status_code != 200:
+            raise SotaError(f"Failed to get the response from {manifest_uri}.")
+        data = json.loads(response.text)
+        logger.debug(f"resp={data}")
+        # Calculate the total size
+        file_size = 0
+        for layer in data['layers']:
+            file_size += layer['size']
+        logger.debug(f"Total file size: {file_size}")
+
+    except (TypeError, KeyError, json.JSONDecodeError, SotaError) as err:
+        err_msg = f"Error getting artifact size from {manifest_uri} using token={jwt_token} Error: {err}"
+        logger.error(err_msg)
+        raise SotaError(err_msg)
+
+    if destination_repo.exists():
+        get_free_space = destination_repo.get_free_space()
+        free_space: int = int(get_free_space)
+    else:
+        raise SotaError("Repository does not exist : " +
+                                  destination_repo.get_repo_path())
+
+    logger.debug("get_free_space: " + repr(get_free_space))
+    logger.debug("Free space available on destination_repo is " + repr(free_space))
+    logger.debug("Free space needed on destination repo is " + repr(file_size))
+    return True if free_space > file_size else False
+
+def parse_uri(uri: CanonicalUri) -> Tuple[str, str, str, str, str, str, str]:
```

Comment:
(see suggested new implementation in other review comment)

Reply:
Added ParsedURI dataclass and updated the function.